x/ibc-transfer/keeper/relay_test.go

```diff
@@ -1,177 +1,260 @@
+				suite.Require().Error(err)
 			}
 		})
 	}
 }
 
+// test receiving coins on chainB with coins that orignate on chainA and
+// coins that orignated on chainB. Coins from source (chainA) have channelB
+// as the denom prefix. The bulk of the testing occurs in the test case
+// for loop since setup is intensive for all cases. The malleate function
+// allows for testing invalid cases.
 func (suite *KeeperTestSuite) TestOnRecvPacket() {
-	data := types.NewFungibleTokenPacketData(prefixCoins2, testAddr1.String(), testAddr2.String())
+	var (
+		channelA, channelB ibctesting.TestChannel
+		coins              sdk.Coins
+		receiver           string
+	)
 
 	testCases := []struct {
 		msg      string
 		malleate func()
+		source   bool
 		expPass  bool
 	}{
-		{"success receive from source chain",
-			func() {}, true},
+		{"success receive from source chain", func() {}, true, true},
+		{"success receive with coins orignated on this chain", func() {}, false, true},
+		{"empty amount", func() {
+			coins = nil
+		}, true, false},
+		{"invalid receiver address", func() {
+			receiver = "gaia1scqhwpgsmr6vmztaa7suurfl52my6nd2kmrudl"
+		}, true, false},
+		{"no dest prefix on coin denom", func() {
+			coins = ibctesting.NewTransferCoins(channelB, "bitcoin", 100)
+		}, false, false},
+
 		// onRecvPacket
-		// - source chain
-		{"no dest prefix on coin denom",
-			func() {
-				data.Amount = testCoins
-			}, false},
-		{"mint failed",
-			func() {
-				data.Amount = prefixCoins2
-				data.Amount[0].Amount = sdk.ZeroInt()
-			}, false},
-		// - receiving chain
-		{"incorrect dest prefix on coin denom",
-			func() {
-				data.Amount = prefixCoins
-			}, false},
-		{"success receive from external chain",
-			func() {
-				data.Amount = prefixCoins
-				escrow := types.GetEscrowAddress(testPort2, testChannel2)
-				_, err := suite.chainA.App.BankKeeper.AddCoins(suite.chainA.GetContext(), escrow, testCoins)
-				suite.Require().NoError(err)
-			}, true},
+		// - coins from source chain (chainA)
+		{"failure: mint zero coins", func() {
+			coins = ibctesting.NewTransferCoins(channelB, sdk.DefaultBondDenom, 0)
+		}, true, false},
+
+		// - coins being sent back to original chain (chainB)
+		{"tries to unescrow more tokens than allowed", func() {
+			coins = ibctesting.NewTransferCoins(channelA, sdk.DefaultBondDenom, 10000)
+		}, false, false},
 	}
 
-	packet := channeltypes.NewPacket(data.GetBytes(), 1, testPort1, testChannel1, testPort2, testChannel2, 100, 0)
-
-	for i, tc := range testCases {
+	for _, tc := range testCases {
 		tc := tc
-		i := i
+
 		suite.Run(fmt.Sprintf("Case %s", tc.msg), func() {
 			suite.SetupTest() // reset
+			_, _, _, _, channelA, channelB = suite.coordinator.Setup(suite.chainA, suite.chainB)
+			receiver = suite.chainB.SenderAccount.GetAddress().String() // must be explicitly changed in malleate
+
+			seq := uint64(1)
+
+			if !tc.source {
+				// send coins from chainB to chainA, receive them, acknowledge them, and send back to chainB
```

Comment:
recv and ack should become a helper func in ibctesting for relaying packets since they are outside the scope of application modules, just need to be passed the correct packet data and ack, or could possibly retrieve the ack itself using the packet sequence. Noted in the testing pkg wishlist #6509